File: leetcode/minimum-size-subarray-sum.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
 public:
  int minSubArrayLen(int target, vector<int>& nums) {
    int left = 0, right = 0;
    int sum = 0, ans = 1000000;
    while (right < nums.size()) {
      sum += nums[right];
      while (sum >= target) {
        if (right - left + 1 < ans) {
          ans = right - left + 1;
        }
        sum -= nums[left];
        left++;
      }
      right++;
    }
    return ans == 1000000 ? 0 : ans;
  };
};

int main() {
  Solution sol = Solution();
  vector<int> nums(6);
  nums.push_back(2);
  nums.push_back(3);
  nums.push_back(1);
  nums.push_back(2);
  nums.push_back(4);
  nums.push_back(3);

  int answer = sol.minSubArrayLen(7, nums);
  cout << answer << endl;
}

File: leetcode/minimum-size-subarray-sum.go
package main

func minSubArrayLen(target int, nums []int) int {
	opts := len(nums) + 1
	sum := 0
	left := 0
	right := 0
	for right = 0; right < len(nums); right++ {
		sum += nums[right]
		for sum >= target {
			if right-left+1 < opts {
				opts = right - left + 1
			}
			sum -= nums[left]
			left += 1
		}
	}
	if opts == len(nums)+1 {
		return 0
	}
	return opts
}
func tc1() {

	numbers := []int{
		2, 3, 1, 2, 4, 3,
	}
	sol := minSubArrayLen(7, numbers)
	println(sol)
}
func tc2() {

	numbers := []int{
		1, 4, 4}
	sol := minSubArrayLen(4, numbers)
	println(sol)
}
func tc3() {

	numbers := []int{
		1, 1, 1, 1, 1, 1, 1, 1,
	}
	sol := minSubArrayLen(11, numbers)
	println(sol)
}

func tc4() {

	numbers := []int{
		1, 2, 3, 4, 5,
	}
	sol := minSubArrayLen(11, numbers)
	println(sol)
}
func main() {
	tc1()
	tc2()
	tc3()
	tc4()
}
